packages/transaction-controller/src/utils/simulation.ts

```diff
@@ -0,0 +1,435 @@
+  const erc721Interface = new Interface(abiERC721);
+  const erc1155Interface = new Interface(abiERC1155);
+
+  return logs
+    .map((currentLog) => {
+      const event = parseLog(
+        currentLog,
+        erc20Interface,
+        erc721Interface,
+        erc1155Interface,
+      );
+
+      if (!event) {
+        log('Failed to parse log', currentLog);
+        return undefined;
+      }
+
+      /* istanbul ignore next */
+      const inputs = event.abi.find((e) => e.name === event.name)?.inputs;
+
+      /* istanbul ignore if */
+      if (!inputs) {
+        log('Failed to find inputs for event', event);
+        return undefined;
+      }
+
+      const args = parseEventArgs(event.args, inputs);
+
+      return {
+        contractAddress: currentLog.address,
+        tokenStandard: event.standard,
+        name: event.name,
+        data: args,
+        abi: event.abi,
+      };
+    })
+    .filter((e) => e !== undefined) as ParsedEvent[];
+}
+
+/**
+ * Parse event arguments using ABI input definitions.
+ * @param args - The raw event arguments.
+ * @param abiInputs - The ABI input definitions.
+ * @returns The parsed event arguments.
+ */
+function parseEventArgs(args: Result, abiInputs: { name: string }[]) {
+  return args.reduce((result, arg, index) => {
+    const name = abiInputs[index].name.replace('_', '');
+    const value = parseEventArgValue(arg);
+
+    result[name] = value;
+
+    return result;
+  }, {});
+}
+
+/**
+ * Parse an event argument value.
+ * @param value - The event argument value.
+ * @returns The parsed event argument value.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function parseEventArgValue(value: any): any {
+  if (Array.isArray(value)) {
+    return value.map(parseEventArgValue);
+  }
+
+  let parsedValue = value;
+
+  if (parsedValue.toHexString) {
+    parsedValue = value.toHexString();
+  }
+
+  return parsedValue.toLowerCase();
+}
+
+/**
+ * Generate token balance changes from parsed events.
+ * @param request - The transaction that was simulated.
+ * @param events - The parsed events.
+ * @returns An array of token balance changes.
+ */
+async function getTokenBalanceChanges(
+  request: GetSimulationDataRequest,
+  events: ParsedEvent[],
+): Promise<SimulationTokenBalanceChange[]> {
+  const balanceTransactionsByToken = getTokenBalanceTransactions(
+    request,
+    events,
+  );
+
+  const balanceTransactions = [...balanceTransactionsByToken.values()];
+
+  log('Generated balance transactions', balanceTransactions);
+
+  if (!balanceTransactions.length) {
+    return [];
+  }
+
+  const response = await simulateTransactions(request.chainId as Hex, {
+    transactions: [...balanceTransactions, request, ...balanceTransactions],
+  });
+
+  if (response.transactions.length !== balanceTransactions.length * 2 + 1) {
+    throw new Error('Invalid response from simulation API');
+  }
+
+  log('Balance simulation response', response);
```

Comment:
Do you want to log the response before throwing the error on line 232, in case it is helpful to debug that error?

Reply:
Done.